09-graphs-p2/02-dijkstra.md

```diff
@@ -1,74 +1,587 @@
+
+```
+current: F
+
+visited nodes: A, B, D, E, C, F
+
+costs/distances:
+    - A: 0
+    - B: 8
+    - C: 6
+    - D: 3
+    - E: 6
+    - F: 8
+```
+<!-- available callout types: info, success, warning, danger, secondary, star  -->
+### !callout-warning
+
+## Limitations of Dijkstra's
+
+There are limitations to Dijkstra's algorithm. Dijkstra's _does not work_ for graphs with negative weights. To find the shortest path in a graph with negative weights we can use the [Bellman Ford algorithm](https://www.programiz.com/dsa/bellman-ford-algorithm). This algorithm is outside the scope of our curriculum and appears only very occasionally in interviews for SDE1 positions, but we encourage you to follow your curiosity.
+
+
+### !end-callout
+
+
+
+## Pseudocode
+The process outlined above can be generalized as pseudocode.
+
+Before jumping into the pseudocode, observe that to find the cost of a path from Node A to a non-neighboring node D, we needed to know the cost of travelling from Node A to Node D's direct neighbor along that path.
+
+In the example below, say we want to find the cost of the path from Node A to Node D. Node D's neighbor along that path is Node C. We could also say that Node C is the _previous_ node along that path. 
+
+Observe that the cost to get from Node A to Node D is the cost to get from Node A to Node C plus the cost to get from Node D to Node C. 
+
+![Previous Node in a Graph Example](./images/dijkstra-subpath.png)
+
+This observation illustrates that when calculating the costs of a path between two non-neighboring nodes as we may need to do in Dijkstra's requires us to know the previous node of the end node in the path that we are looking at. As a result, when implementing Djikstra's algorithm it can be useful for us to keep track of each node's previous node along its minimum cost path.
+
+
+The following pseudocode implementation of Dijkstra's assumes that we are provided with a graph and a starting node.
+```
+- Create a distances list equal in length to the number of nodes in the graph
+    - Tracks the minimum cost/distance from start node to each node in graph
+    - Initialize each value to infinity
+    - Initialize distances[start_node] to zero
+- Create a previous list equal in length to the number of nodes in the graph
+    - Tracks the previous node in each node's shortest/minimum cost path
+    - Initialize each value to None since we have not yet traversed any paths
+- Create a visited list to track which nodes have already been visited
+- Create a priority queue and add the start node to it
+- While the queue is not empty:
+    - Set current as node in the queue with minimum cost
+    - Add current to the list of visited nodes
+    - loop through all the current node's neighbors
+        - if the neighbor has not yet been visited:
+            - set that node to visited
```

Comment:
I took a crack at writing this by following the pseudocode. This walkthrough is amazing!! The one thing that tripped me up is the line that says "set that node to visited" in the loop to visit the current node's neighbors. I took that to mean "mark the node as visited", which led to an incorrect result. 

I dunno if I was just misreading it, but there might be an opportunity to update the wording here to be less ambiguous? I might even just strip it out completely.

Reply:
Oh this is definitely a mistake. This should be taken out, nice catch!